pkg/internal/server/grpc/validators.go

```diff
@@ -17,42 +17,101 @@ limitations under the License.
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSecurityTokenMissing)
 	}
+
 	if len(req.GetNamespace()) == 0 {
-		return status.Errorf(codes.InvalidArgument, msgInvalidArgumentNamespaceMissing)
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentNamespaceMissing)
 	}
+
 	if len(req.GetBaseSnapshotName()) == 0 {
-		return status.Errorf(codes.InvalidArgument, msgInvalidArgumentBaseSnapshotNameMissing)
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentBaseSnapshotNameMissing)
 	}
+
 	if len(req.GetTargetSnapshotName()) == 0 {
-		return status.Errorf(codes.InvalidArgument, msgInvalidArgumentTargetSnapshotNameMissing)
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentTargetSnapshotNameMissing)
+	}
+
+	baseSnapshotHandle, driver, err := getVolSnapshotInfo(ctx, s.snapshotClient(), req.Namespace, req.BaseSnapshotName)
+	if err != nil {
+		return "", "", err
 	}
 
-	return nil
+	if driver != s.driverName() {
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSnaphotDriverInvalidFmt, s.driverName())
+	}
+
+	targetSnapshotHandle, driver, err := getVolSnapshotInfo(ctx, s.snapshotClient(), req.Namespace, req.TargetSnapshotName)
+	if err != nil {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableVolumeSnapshotNotReady)
+	}
+
+	if driver != s.driverName() {
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSnaphotDriverInvalidFmt, s.driverName())
+	}
+
+	return baseSnapshotHandle, targetSnapshotHandle, nil
+}
+
+// getVolSnapshotInfo returns snapshot handle and csi driver info of the VolumeSnapshot.
+func getVolSnapshotInfo(ctx context.Context, cli snapshot.Interface, namespace, vsName string) (string, string, error) {
+	vs, err := cli.SnapshotV1().VolumeSnapshots(namespace).Get(ctx, vsName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableFailedToGetVolumeSnapshotFmt, err)
+	}
+	if vs.Status.ReadyToUse == nil || !*vs.Status.ReadyToUse {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableVolumeSnapshotNotReadyFmt, vsName)
+	}
+	if vs.Status.BoundVolumeSnapshotContentName == nil {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableInvalidVolumeSnapshotStatusFmt, vsName)
+	}
+
+	vsc, err1 := cli.SnapshotV1().VolumeSnapshotContents().Get(ctx, *vs.Status.BoundVolumeSnapshotContentName, metav1.GetOptions{})
+	if err1 != nil {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableFailedToGetVolumeSnapshotContentFmt, err)
+	}
+	if vsc.Status.ReadyToUse == nil || !*vsc.Status.ReadyToUse {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableVolumeSnapshotContentNotReadyFmt, vsc.Name)
+	}
+	if vsc.Status.SnapshotHandle == nil {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableInvalidVolumeSnapshotContentStatusFmt, vsName)
+	}
+	return *vsc.Status.SnapshotHandle, vsc.Spec.Driver, nil
 }
```

Comment:
I think this should go into perhaps a `utils.go` file. Future support to load `Secrets` would also reside there.